src/hotspot/cpu/x86/c2_stubGenerator_x86_64_string.cpp

```diff
@@ -0,0 +1,1837 @@
+
+  // Check for trivial cases
+  // needle length == 0?
+  __ cmpq(needle_len_p, 0);
+  __ jg_b(L_nextCheck);
+  __ xorl(rax, rax);
+  __ leave();
+  __ ret(0);
+
+  __ bind(L_nextCheck);
+  // haystack length >= needle length?
+  __ movq(rax, haystack_len_p);
+  __ subq(rax, needle_len_p);
+  __ jge_b(L_checksPassed);
+
+  __ movq(rax, -1);
+  __ leave();
+  __ ret(0);
+
+  __ bind(L_checksPassed);
+
+  // Check for highly-optimized ability - haystack <= 32 bytes and needle <= 6 bytes
+  // haystack_len is in elements, not bytes, for UTF-16
+  __ cmpq(haystack_len_p, isU ? 16 : 32);
+  __ ja(L_begin);
+
+  // needle_len is in elements, not bytes, for UTF-16 <=> UTF-16
+  __ cmpq(needle_len_p, isUU ? 3 : 6);
+  __ ja(L_begin);
+
+  // Handle short haystack and needle specially
+  // Generated code does not return - either found or not
+  highly_optimized_short_cases(ae, haystack_p, haystack_len_p, needle_p, needle_len_p, xmm0, xmm1,
+                               r10, r11, _masm);
+
+  // If we're generating UL, we need to "pretend" we're generating UU code
+  // for the case where the needle can be expanded onto the stack
+  if (isReallyUL) {
+    ae = StrIntrinsicNode::UU;
+    isUL = false;
+    isUU = true;
+  }
+
+  // Set up jump tables.  Used when needle size <= NUMBER_OF_CASES
+  setup_jump_tables(ae, L_returnError, L_returnR11, L_bigCaseFixupAndReturn, &big_jump_table,
+                    &small_jump_table, _masm);
+
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  ////////////////////////////////////////////////////////////////////////////////////////
+  //
+  // The above code handles all cases (LL, UL, UU) for haystack size <= 32 bytes
+  // and needle size <= 6 bytes.
+  //
+  // Main processing proceeds as follows:
+  //  Save state and setup stack, etc.
+  //  If UL, jump to code to handle special-case UL situations (see L_wcharBegin below)
+  //  Broadcast the first and last needle elements to XMM_BYTE_0 and XMM_BYTE_K, respectively
+  //  If the length in bytes of the haystack is > 32, dispatch to the big switch handling code
+  //  If the haystack length in bytes is <= 32:
+  //    Copy the haystack to the stack.  This is done to prevent possible page faults and
+  //      allows for reading full 32-byte chunks of the haystack.
+  //    Dispatch to the small switch handling code
+  //
+  // Here, "big switch" and "small switch" refers to the haystack size: > 32 bytes for big
+  // and <= 32 bytes for small.  The switches implement optimized code for handling 1 to
+  // NUMBER_OF_CASES (currently 10) needle sizes for both big and small.  There are special
+  // routines for handling needle sizes > NUMBER_OF_CASES (L_{big,small}CaseDefault).  These
+  // cases use C@'s arrays_equals() to compare the needle to the haystack.  The small cases
```

Comment:
```suggestion
  // cases use C2's arrays_equals() to compare the needle to the haystack.  The small cases
```
Randomly spotted this.

Reply:
Fixed.